[Page color sampling] Refactor fixed container edge sampling so that we keep track of state on all rect edges

whsieh opened the pull request:
#### fa24f4dd4063c526c78440b381fce8baee934a5e
<pre>
[Page color sampling] Refactor fixed container edge sampling so that we keep track of state on all rect edges
<a href="https://bugs.webkit.org/show_bug.cgi?id=293165">https://bugs.webkit.org/show_bug.cgi?id=293165</a>

Reviewed by Abrar Rahman Protyasha.

Part 1 of refactoring the fixed container edge sampling heuristic to be more stable — this adds a
member to `Page` to store the last `FixedContainerEdges` state. Instead of computing this and
sending it along in the remote layer tree transaction, we now call `updateFixedContainerEdges` and
send along the current state.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
(WebCore::Page::updateFixedContainerEdges):
(WebCore::Page::lastTopFixedContainerColor const):
* Source/WebCore/page/Page.h:

Replace `m_lastTopFixedContainerColor` (which only keeps track of the last predominant color on the
top edge) with `m_fixedContainerEdges`, which keeps track of all sampled edges. Note that this is a
`UniqueRef`, in order to avoid including `FixedContainerEdges.h` in `Page.h`.

(WebCore::Page::fixedContainerEdges const):
(WebCore::Page::setLastTopFixedContainerColor): Deleted.
(WebCore::Page::lastTopFixedContainerColor const): Deleted.
* Source/WebCore/platform/FixedContainerEdges.cpp: Copied from Source/WebCore/platform/FixedContainerEdges.h.

Add an implementation file for `FixedContainerEdges` so that it can contain the
`WTF_MAKE_TZONE_ALLOCATED_IMPL`. Also move a couple of inline method definitions to the source file,
to make the header somewhat more readable.

(WebCore::FixedContainerEdges::hasFixedEdge const):
(WebCore::FixedContainerEdges::predominantColor const):
* Source/WebCore/platform/FixedContainerEdges.h:
(WebCore::FixedContainerEdges::FixedContainerEdges):
(WebCore::FixedContainerEdges::hasFixedEdge const): Deleted.
(WebCore::FixedContainerEdges::predominantColor const): Deleted.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.cpp:

Unified source build fixes (include missing headers).

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Canonical link: <a href="https://commits.webkit.org/295058@main">https://commits.webkit.org/295058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c22b1f06c693a8355c403b4af9bc725da1dc0b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109171 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54640 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32224 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78986 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18667 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93794 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59314 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18457 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11845 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54004 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88196 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11903 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111556 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31132 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22940 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87998 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89994 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87655 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22306 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32539 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10281 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25522 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31061 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36371 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30854 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34191 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32415 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4964,9 +4964,8 @@ void WebPage::willCommitLayerTree(RemoteLayerTreeTransaction& layerTransaction,
-        protectedCorePage()->setLastTopFixedContainerColor(fixedContainerEdges.predominantColor(BoxSide::Top));
-        layerTransaction.setFixedContainerEdges(WTFMove(fixedContainerEdges));
+        page->updateFixedContainerEdges(sidesRequiringFixedContainerEdges());
+        layerTransaction.setFixedContainerEdges(FixedContainerEdges { page->fixedContainerEdges() });
```
You can probably skip the type hint because the copy ctor isn't explicit

whsieh replied:
Oh, I got an error with that approach:

```
In file included from […]/Build/Release/DerivedSources/WebKit/unified-sources/UnifiedSource137.cpp:2:
[…]/Source/WebKit/WebProcess/WebPage/WebPage.cpp:4968:49: error: binding reference of type 'WebCore::FixedContainerEdges' to value of type 'const FixedContainerEdges' drops 'const' qualifier
 4968 |         layerTransaction.setFixedContainerEdges({ page->fixedContainerEdges() });
In file included from […]/Build/Release/DerivedSources/WebKit/unified-sources/UnifiedSource137.cpp:2:
```

...but this made me realize that I could just make the parameter type a `const&` instead of `&&` (we end up needing a copy anyways, so this shouldn't be less efficient).

Thanks for the review!

whsieh committed:
[Page color sampling] Refactor fixed container edge sampling so that we keep track of state on all rect edges
https://bugs.webkit.org/show_bug.cgi?id=293165

Reviewed by Abrar Rahman Protyasha.

Part 1 of refactoring the fixed container edge sampling heuristic to be more stable — this adds a
member to `Page` to store the last `FixedContainerEdges` state. Instead of computing this and
sending it along in the remote layer tree transaction, we now call `updateFixedContainerEdges` and
send along the current state.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
(WebCore::Page::updateFixedContainerEdges):
(WebCore::Page::lastTopFixedContainerColor const):
* Source/WebCore/page/Page.h:

Replace `m_lastTopFixedContainerColor` (which only keeps track of the last predominant color on the
top edge) with `m_fixedContainerEdges`, which keeps track of all sampled edges. Note that this is a
`UniqueRef`, in order to avoid including `FixedContainerEdges.h` in `Page.h`.

(WebCore::Page::fixedContainerEdges const):
(WebCore::Page::setLastTopFixedContainerColor): Deleted.
(WebCore::Page::lastTopFixedContainerColor const): Deleted.
* Source/WebCore/platform/FixedContainerEdges.cpp: Copied from Source/WebCore/platform/FixedContainerEdges.h.

Add an implementation file for `FixedContainerEdges` so that it can contain the
`WTF_MAKE_TZONE_ALLOCATED_IMPL`. Also move a couple of inline method definitions to the source file,
to make the header somewhat more readable.

(WebCore::FixedContainerEdges::hasFixedEdge const):
(WebCore::FixedContainerEdges::predominantColor const):
* Source/WebCore/platform/FixedContainerEdges.h:
(WebCore::FixedContainerEdges::FixedContainerEdges):
(WebCore::FixedContainerEdges::hasFixedEdge const): Deleted.
(WebCore::FixedContainerEdges::predominantColor const): Deleted.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.cpp:

Unified source build fixes (include missing headers).

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willCommitLayerTree):

Canonical link: https://commits.webkit.org/295058@main
